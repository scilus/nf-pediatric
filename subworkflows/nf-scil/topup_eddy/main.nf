// ** Importing modules from nf-scil ** //
include { PREPROC_TOPUP } from '../../../modules/nf-scil/preproc/topup/main'
include { PREPROC_EDDY } from '../../../modules/nf-scil/preproc/eddy/main'
include { UTILS_EXTRACTB0 } from '../../../modules/nf-scil/utils/extractb0/main'

workflow TOPUP_EDDY {

    // ** The subworkflow will optionally run topup if a reverse b0 or reverse DWI is provided.   ** //
    // ** In both cases, it will perform EDDY and also extract a b0 from the corrected DWI image. ** //

    take:
        ch_dwi // channel: [ val(meta), [ dwi, bval, bvec ]
        ch_b0 // channel: [ val(meta), b0 ]
        ch_rev_dwi // channel: [ val(meta), [ rev_dwi, rev_bval, rev_bvec ]
        ch_rev_b0 // channel: [ val(meta), rev_b0Â ]
        ch_config_topup // channel

    main:
        ch_versions = Channel.empty()

        // ** Create channel for TOPUP ** //
        if ( ch_rev_dwi )
        {
            ch_image =    ch_dwi.join(ch_b0)
                                .join(ch_rev_dwi)
            ch_eddy_input =   ch_dwi.combine(ch_rev_dwi, by: 0)
        }
        else {
            ch_image =    ch_dwi.join(ch_b0)
                                .map{ it + [[], [], []] }
            ch_eddy_input =   ch_dwi.map{ it + [[], [], []] }
        }
        if ( ch_rev_b0 )
        {
            ch_image =    ch_image.join(ch_rev_b0)
        }
        else {
            ch_image =    ch_image.map{ it + [[]] }
        }

        if ( ch_rev_dwi || ch_rev_b0 )
        {
            // ** RUN TOPUP ** //
            PREPROC_TOPUP ( ch_image, ch_config_topup )
            ch_versions = ch_versions.mix(PREPROC_TOPUP.out.versions.first())

            // ** Create channel for EDDY ** //
            ch_eddy_input =    ch_eddy_input.combine(PREPROC_TOPUP.out.topup_corrected_b0s, by: 0)
                                            .combine(PREPROC_TOPUP.out.topup_fieldcoef, by: 0)
                                            .combine(PREPROC_TOPUP.out.topup_movpart, by: 0)
        }
        else
        {
            // ** RUN EDDY ** //
            ch_eddy_input =    ch_dwi.map{ it + [[], [], [], [], [], []] }
        }

        PREPROC_EDDY ( ch_eddy_input )
        ch_dwi_extract_b0 =   PREPROC_EDDY.out.dwi_corrected.combine(PREPROC_EDDY.out.bval_corrected, by: 0)
                                                            .combine(PREPROC_EDDY.out.bvec_corrected, by: 0)
        UTILS_EXTRACTB0 { ch_dwi_extract_b0 }

        ch_versions = ch_versions.mix(UTILS_EXTRACTB0.out.versions.first())
        ch_versions = ch_versions.mix(PREPROC_EDDY.out.versions.first())

    emit:
        dwi      = PREPROC_EDDY.out.dwi_corrected       // channel: [ val(meta), [ dwi_corrected ] ]
        bval     = PREPROC_EDDY.out.bval_corrected      // channel: [ val(meta), [ bval_corrected ] ]
        bvec     = PREPROC_EDDY.out.bvec_corrected      // channel: [ val(meta), [ bvec_corrected ] ]
        b0       = UTILS_EXTRACTB0.out.b0               // channel: [ val(meta), [ b0 ] ]
        b0_mask  = PREPROC_EDDY.out.b0_mask             // channel: [ val(meta), [ b0_mask ] ]
        versions = ch_versions                  // channel: [ versions.yml ]
}
