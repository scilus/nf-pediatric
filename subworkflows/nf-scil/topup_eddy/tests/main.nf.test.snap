{
    "topup_eddy_rev_b0": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__b0_bet_mask.nii.gz:md5,fc1f94e2a5e2cf5197d7fdc28a83ca28"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bval_eddy:md5,8192e97b08e8032382397bb844b31892"
                ]
            ],
            [
                "versions.yml:md5,1210f8705c10f136e83e5f07c0d77b70",
                "versions.yml:md5,3e57e2f1a94bbabf9be75f50b4cbcdb0",
                "versions.yml:md5,dc9478b4d3cb93d93799e493aefe5220"
            ],
            "test__dwi_corrected.nii.gz",
            "test__dwi_eddy_corrected.bvec",
            "test_b0.nii.gz"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-09T00:17:01.349981"
    },
    "eddy": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__b0_bet_mask.nii.gz:md5,3574cc8f29da429a0f610f2572c4886c"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bval_eddy:md5,8192e97b08e8032382397bb844b31892"
                ]
            ],
            [
                "versions.yml:md5,1210f8705c10f136e83e5f07c0d77b70",
                "versions.yml:md5,dc9478b4d3cb93d93799e493aefe5220"
            ],
            "test__dwi_corrected.nii.gz",
            "test__dwi_eddy_corrected.bvec",
            "test_b0.nii.gz"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-09T00:15:24.300649"
    },
    "topup_eddy_rev_dwi": {
        "content": [
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__b0_bet_mask.nii.gz:md5,fc1f94e2a5e2cf5197d7fdc28a83ca28"
                ]
            ],
            [
                [
                    {
                        "id": "test",
                        "single_end": false
                    },
                    "test__bval_eddy:md5,4c61c53078316c31b4d5daf446a3d6ac"
                ]
            ],
            [
                "versions.yml:md5,1210f8705c10f136e83e5f07c0d77b70",
                "versions.yml:md5,3e57e2f1a94bbabf9be75f50b4cbcdb0",
                "versions.yml:md5,dc9478b4d3cb93d93799e493aefe5220"
            ],
            "test__dwi_corrected.nii.gz",
            "test__dwi_eddy_corrected.bvec",
            "test_b0.nii.gz"
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-09T00:19:03.319712"
    }
}